File: agent/ocop.py
# -*- coding: utf-8 -*-
"""Rút hạng sao OCOP từ entity — MỘT nơi duy nhất cho backend.

Song sinh với `web-nuxt/utils/ocop.ts`. Hai bản là BẮT BUỘC hôm nay vì backend
dựng JSON-LD, văn bản nạp cho LLM và thẻ chat, còn frontend dựng huy hiệu — mà
API chưa phát ra trường hạng đã chuẩn hoá. Đường thoát đúng là backend chiếu sẵn
một trường hạng để frontend khỏi tự rút; đã ghi ROADMAP.

Sửa luật ở đây thì PHẢI sửa cả bản kia. Hai bộ test soi CÙNG MỘT bộ chuỗi thật
(`agent/tests/test_ocop.py` và `web-nuxt/tests/ocop-stars.test.ts`) nên chúng sẽ
cùng đỏ nếu lệch.
"""
from __future__ import annotations

import re
from typing import Any

_RE_OCOP_SELF_TIER = re.compile(r"^\s*(?:ocop\s*)?([1-5])\s*sao\b", re.IGNORECASE)
_RE_OCOP_PROPOSAL = re.compile(
    r"đề xuất|đề nghị|chờ\s+(?:công nhận|đánh giá|xét)|đang\s+(?:xét|đề nghị)|dự kiến",
    re.IGNORECASE)
_RE_OCOP_AWARDED = re.compile(
    r"đạt|được công nhận|đã công nhận|chứng nhận|cấp quốc gia", re.IGNORECASE)
_OCOP_CLAIM_WINDOW = 40
_OCOP_NUMERIC_KEYS = ("ocop_star", "ocop_stars", "ocop_rating")

# Chương trình OCOP chỉ công nhận 3, 4 và 5 sao. KHÔNG có "OCOP 1 sao" hay
# "OCOP 2 sao" — chúng không tồn tại, nên một ô số mang 1 hoặc 2 không phải hạng.
_OCOP_MIN_GRADE = 3


def _attrs(entity: dict[str, Any]) -> dict[str, Any]:
    """`attributes` đã chuẩn hoá về dict.

    Dữ liệu thật có entity mang `attributes` là LIST (dị dạng) — bản đầu nổ
    AttributeError ở đó. Một test cũ của seo bơm đúng ca ấy và đã bắt được; giữ
    chốt chặn ở tầng thấp nhất thay vì rải ở từng nơi gọi.
    """
    a = entity.get("attributes") or {}
    return a if isinstance(a, dict) else {}


def _tu_khoa_so(attrs: dict[str, Any]) -> int:
    """Hạng từ ba khoá SỐ. `bool` bị loại tường minh: `True` là `1` trong Python
    nên không chặn thì `ocop_star: true` thành "OCOP 1 sao"."""
    for key in _OCOP_NUMERIC_KEYS:
        raw = attrs.get(key)
        if isinstance(raw, bool):
            continue
        if isinstance(raw, (int, float)):
            return int(raw)
        if isinstance(raw, str) and raw.strip().isdigit() and 1 <= int(raw.strip()) <= 5:
            return int(raw.strip())
    return 0


def _tu_o_ocop(attrs: dict[str, Any]) -> int:
    """Hạng từ ô `ocop` — văn xuôi tự do, nên NEO ĐẦU CHUỖI có chủ đích.

    Nới thành "tìm N sao ở bất kỳ đâu" là tự phong 5 sao cho trái dừa bằng danh
    mục sản phẩm của VICOSAP. Số nguyên thì rõ nghĩa nên nhận thẳng (bản TS cũng
    nhận; dữ liệu thật hiện không có ca này nhưng nó là hình dạng hợp lệ).
    """
    text = attrs.get("ocop")
    if isinstance(text, bool):
        return 0
    if isinstance(text, (int, float)):
        return int(text)
    if isinstance(text, str):
        m = _RE_OCOP_SELF_TIER.match(text)
        if m:
            return int(m.group(1))
    return 0


def _claimed_tier(attrs: dict[str, Any]) -> int:
    """Hạng GHI TRONG DỮ LIỆU, chưa qua bộ lọc §1.7. Khoá số thắng văn xuôi."""
    return _tu_khoa_so(attrs) or _tu_o_ocop(attrs)


def _van_xuoi_xac_nhan_hang(entity: dict[str, Any], tier: int) -> bool:
    """Văn xuôi có XÁC NHẬN hạng này cho CHÍNH entity không?

    Dùng lại đúng cửa sổ ±40 ký tự quanh "N sao" của luật §1.7 bên dưới, nên hai
    chỗ không thể nói ngược nhau.
    """
    prose = f"{entity.get('summary') or ''} {entity.get('description') or ''}"
    for m in re.finditer(rf"{tier}\s*sao", prose, re.IGNORECASE):
        window = prose[max(0, m.start() - _OCOP_CLAIM_WINDOW):m.end() + _OCOP_CLAIM_WINDOW]
        if _RE_OCOP_AWARDED.search(window):
            return True
    return False


def _o_so_khong_phai_tin_hieu_ocop(entity: dict[str, Any], attrs: dict[str, Any]) -> bool:
    """Con số trong ô OCOP là RÁC CHÉP NHẦM CỘT, không phải một chứng nhận.

    Đo trên `web/data.json` 2026-08-30 — 13 cơ sở lưu trú mang `ocop_star`, và
    CẢ 13 có `ocop_star` bằng đúng `star_rating` của chính nó (1=1, 2=2, 4=4,
    5=5), `ocop_certified` rỗng. Đó là vân tay của một lượt nhập chép nhầm cột
    hạng-sao-khách-sạn sang ô OCOP. Hậu quả: trang chi tiết in "Sản phẩm OCOP
    1 sao — Chương trình Mỗi xã Một sản phẩm" cho một khách sạn. Khai khống một
    chứng nhận NHÀ NƯỚC là đúng thứ CLAUDE.md §1.7 cấm.

    HAI LUẬT, mỗi luật tự đứng được:

    (A) Hạng dưới 3 — OCOP không có bậc đó. Bắt 12 entity (11 khách sạn +
        `khu-du-lich-truong-an`); trong toàn kho KHÔNG có sản phẩm nào mang
        hạng 1|2, nên luật này không đụng một sản phẩm thật nào.

    (B) Số chỉ chép lại `star_rating` VÀ văn xuôi không xác nhận hạng đó. Bắt
        nốt `homestay-sokfram` (5=5, văn xuôi chỉ nói nó BÀY BÁN sản phẩm OCOP
        3–5 sao của địa phương — cùng bẫy "danh mục của người khác" mà
        `dua-sap-cau-ke` đã dạy).

    VÌ SAO luật (B) phải có vế văn xuôi: `somo-farm-cuu-long` cũng có 4=4 nhưng
    văn xuôi ghi "Đạt chứng nhận OCOP 4 sao năm 2023 cho sản phẩm du lịch sinh
    thái" — OCOP nhóm 6 (dịch vụ du lịch) là CÓ THẬT. Luật (B) thiếu vế đó sẽ
    bóp mất một chứng nhận đúng. Đã đối chiếu từng ca trong cả 13.

    Không đụng `ocop_star: 9` / `0` / `true`: chúng vẫn giữ hành vi cũ (không rút
    được hạng nhưng vẫn tính là có dấu hiệu) vì đó là ý định ghi OCOP kèm lỗi gõ,
    khác hẳn việc chép nhầm nguyên một cột khác.
    """
    n = _tu_khoa_so(attrs)
    if n <= 0:
        return False
    if n < _OCOP_MIN_GRADE:
        return True
    sao_luu_tru = attrs.get("star_rating")
    if sao_luu_tru in (None, ""):
        return False
    try:
        cung_so = int(str(sao_luu_tru).strip()) == n
    except (TypeError, ValueError):
        return False
    return cung_so and not _van_xuoi_xac_nhan_hang(entity, n)


def _co_dau_hieu_ocop(entity: dict[str, Any], attrs: dict[str, Any]) -> bool:
    if (any(attrs.get(k) not in (None, "") for k in _OCOP_NUMERIC_KEYS)
            and not _o_so_khong_phai_tin_hieu_ocop(entity, attrs)):
        return True
    return bool(attrs.get("ocop") or attrs.get("ocop_certified"))


def _ocop_tier_is_provisional(entity: dict[str, Any], tier: int) -> bool:
    """Hạng ghi trong dữ liệu có phải hạng ĐÃ ĐẠT, hay mới chỉ được đề nghị?

    Luật HAI CHIỀU. Luật một chiều ("văn xuôi có chữ đề xuất thì hạ hạng") đánh
    oan ngay ca thật đầu tiên: `khoai-lang-say-dong-phat` ghi "ĐẠT OCOP 4 sao và
    được đề xuất công nhận 5 sao" — hạng 4 của nó CÓ THẬT. Nên chỉ hạ khi cửa sổ
    có lời đề nghị mà KHÔNG có lời xác nhận.

    Xét TOÀN VĂN chứ không từng lần nhắc: `khoai-lang-say-binh-tan` nói "được đề
    xuất lên Trung ương đánh giá 5 sao" hai lần rồi kết bằng câu trần "Sản phẩm
    OCOP 5 sao." Luật xét-từng-lần sẽ để câu trần đó lật ngược cả phán quyết.
    """
    prose = f"{entity.get('summary') or ''} {entity.get('description') or ''}"
    if not prose.strip():
        return False
    saw_proposal = saw_awarded = False
    for m in re.finditer(rf"{tier}\s*sao", prose, re.IGNORECASE):
        window = prose[max(0, m.start() - _OCOP_CLAIM_WINDOW):m.end() + _OCOP_CLAIM_WINDOW]
        if _RE_OCOP_PROPOSAL.search(window):
            saw_proposal = True
        if _RE_OCOP_AWARDED.search(window):
            saw_awarded = True
    return saw_proposal and not saw_awarded


def ocop_display_label(entity: dict[str, Any]) -> str:
    """«OCOP 5 sao» · «OCOP» · '' — KHÔNG BAO GIỜ là văn xuôi thô.

    `attributes.ocop` là chuỗi tự do. Trước bản vá này JSON-LD phát thẳng nó, đo
    được trên trang đang chạy 2026-08-27: `dua-sap-cau-ke` có
        brand.name = "OCOP VICOSAP: 4 SP OCOP 5 sao quốc gia + 7 SP OCOP 4 sao"
    tức khai với máy tìm kiếm rằng thương hiệu của TRÁI DỪA là danh mục chứng
    nhận của một CÔNG TY KHÁC. Một entity khác mang cả số quyết định:
        "OCOP 4 sao (QĐ 114/QĐ-UBND, 15/1/2020)"

    NỢ SONG BẢN — ghi rõ để không ai tưởng là trùng lặp vô ý: luật này còn một
    bản TypeScript ở `web-nuxt/utils/ocop.ts` cho huy hiệu trên giao diện. Hai
    bản là BẮT BUỘC hôm nay vì JSON-LD dựng ở backend còn huy hiệu dựng ở
    frontend, và API chưa phát ra trường hạng đã chuẩn hoá. Đường thoát đúng là
    backend chiếu sẵn một trường hạng để frontend khỏi tự rút — task riêng, đã
    ghi ROADMAP. Sửa luật ở đây thì PHẢI sửa cả bản kia; hai bộ test soi cùng
    một bộ chuỗi thật nên chúng sẽ cùng đỏ nếu lệch.
    """
    attrs = _attrs(entity)
    if _o_so_khong_phai_tin_hieu_ocop(entity, attrs):
        # Ô số là rác chép nhầm cột — bỏ nó, chỉ còn văn xuôi được nói.
        tier = _tu_o_ocop(attrs)
    else:
        tier = _claimed_tier(attrs)
    if tier > 0 and _ocop_tier_is_provisional(entity, tier):
        tier = 0   # §1.7 — hạng mới ĐỀ NGHỊ không phải hạng đã đạt
    if 1 <= tier <= 5:
        return f"OCOP {tier} sao"
    return "OCOP" if _co_dau_hieu_ocop(entity, attrs) else ""


def ocop_tier(entity: dict[str, Any]) -> int:
    r"""Hạng DÙNG ĐƯỢC, 0 = có chứng nhận nhưng chưa rõ hạng (hoặc mới đề nghị).

    Dùng cho bộ lọc "từ N sao trở lên". Bản cũ ở `server.py` rút hạng bằng
    `re.search(r"(\d)")` — bắt CHỮ SỐ ĐẦU TIÊN ở bất kỳ đâu trong chuỗi, nên
    "VICOSAP: 4 SP OCOP 5 sao quốc gia..." cho ra 4, và một chuỗi có năm ban
    hành sẽ cho ra chữ số của năm.
    """
    label = ocop_display_label(entity)
    m = re.match(r"OCOP ([1-5]) sao$", label)
    return int(m.group(1)) if m else 0


def is_ocop_certified(entity: dict[str, Any]) -> bool:
    """Có dấu hiệu OCOP nào không — KỂ CẢ khi không rút được hạng.

    Lọc bằng `attributes.ocop` truthy bỏ sót 73 sản phẩm chỉ mang `ocop_star`
    (đo 2026-08-27). Đó là cùng một lỗi đã vá ở trang /ocop, còn sống trong
    xếp hạng, đếm và bộ lọc tìm kiếm của backend.
    """
    return _co_dau_hieu_ocop(entity, _attrs(entity))

File: web-nuxt/utils/ocop.ts
// Rút hạng sao OCOP từ entity — MỘT nơi duy nhất.
//
// Dữ liệu ghi hạng sao ở BỐN kiểu khoá khác nhau (nợ chuẩn hoá, ROADMAP §31.5):
// `ocop_star` (81 sản phẩm) · `ocop_stars` · `ocop_rating` · `ocop` (văn xuôi tự
// do). Trang /ocop trước đây chỉ đọc `parseInt(attributes.ocop)`, mà `ocop` là
// câu chữ chứ không phải số — nên `parseInt('OCOP 3 sao')` = NaN → 0. Hậu quả đo
// được trên trang đang chạy: sổ vinh danh hiện 3 sản phẩm thay vì 77, dải "Bậc 5
// sao" KHÔNG render, triện son khai hạng 4 trong khi dữ liệu có 5 sản phẩm
// 5 sao.
//
// Chuẩn hoá DỮ LIỆU (gộp 4 khoá về 1) là task riêng cần backup B1 + chỉ đạo chủ
// dự án. Ở đây chỉ sửa BÊN ĐỌC cho chịu được cả bốn kiểu — additive-first (B2),
// không đụng một hàng dữ liệu nào.

export interface OcopEntityLike {
  summary?: string | null
  description?: string | null
  attributes?: Record<string, any> | null | undefined
}

const NUMERIC_KEYS = ['ocop_star', 'ocop_stars', 'ocop_rating'] as const

// NEO ĐẦU CHUỖI có chủ đích. Nới ra thành "tìm N sao ở bất kỳ đâu" sẽ tự phong
// 5 sao cho `dua-sap-cau-ke`, vì `ocop` của nó là
//   "VICOSAP: 4 SP OCOP 5 sao quốc gia + 7 SP OCOP 4 sao"
// — đó là DANH MỤC SẢN PHẨM CỦA MỘT CÔNG TY KHÁC, không phải chứng nhận của trái
// dừa. Neo đầu bắt được 10/11 chuỗi có hạng ("OCOP 3 sao", "4 sao",
// "OCOP 4 sao (QĐ 114/QĐ-UBND, 15/1/2020)"…) và bỏ sót đúng 1 ca lành
// ("Cam sành Khánh Nhân, Thuận Loan — OCOP 3 sao"). Bỏ sót thì sản phẩm hiện là
// "có chứng nhận, chưa rõ hạng" — thiệt thòi nhưng THẬT; bắt nhầm thì trang khai
// khống một hạng quốc gia.
const SELF_TIER = /^\s*(?:ocop\s*)?([1-5])\s*sao\b/i

// §1.7 — hạng "mới được đề nghị" KHÔNG phải hạng đã đạt.
const PROPOSAL = /đề xuất|đề nghị|chờ\s+(?:công nhận|đánh giá|xét)|đang\s+(?:xét|đề nghị)|dự kiến/i
const AWARDED = /đạt|được công nhận|đã công nhận|chứng nhận|cấp quốc gia/i
const CLAIM_WINDOW = 40

// Chương trình OCOP chỉ công nhận 3, 4 và 5 sao. KHÔNG có "OCOP 1 sao" hay
// "OCOP 2 sao" — chúng không tồn tại, nên ô số mang 1 hoặc 2 không phải hạng.
const MIN_GRADE = 3

function attrs(e: OcopEntityLike | null | undefined): Record<string, any> {
  return (e && e.attributes) || {}
}

/** Hạng sao GHI TRONG DỮ LIỆU, chưa qua bộ lọc §1.7. 0 = không rút được hạng. */
/** Hạng từ BA KHOÁ SỐ. Tách ra để luật "ô số là rác chép nhầm cột" soi được
 *  riêng phần số, giống `_tu_khoa_so` bên bản Python. */
function numericKeyTier(a: Record<string, any>): number {
  for (const key of NUMERIC_KEYS) {
    const raw = a[key]
    if (typeof raw === 'number' && Number.isFinite(raw)) return Math.trunc(raw)
    // Chuỗi thuần số ("4") vẫn là con số người ta định ghi.
    if (typeof raw === 'string' && /^\s*[1-5]\s*$/.test(raw)) return parseInt(raw, 10)
  }
  return 0
}

/** Hạng từ Ô `ocop` (văn xuôi tự do), tách riêng như `_tu_o_ocop` bên Python. */
function proseFieldTier(a: Record<string, any>): number {
  const text = a.ocop
  // Số nguyên trong ô `ocop` là RÕ NGHĨA, khác hẳn văn xuôi — nhận thẳng. Bản
  // Python (`agent/ocop.py`) nhận ca này; hai bản sinh đôi phải khớp nhau.
  if (typeof text === 'number' && Number.isFinite(text)) return Math.trunc(text)
  if (typeof text === 'string') {
    const m = SELF_TIER.exec(text)
    if (m) return parseInt(m[1]!, 10)
  }
  return 0
}

export function ocopClaimedStars(e: OcopEntityLike | null | undefined): number {
  const a = attrs(e)
  return numericKeyTier(a) || proseFieldTier(a)
}

/** Văn xuôi có XÁC NHẬN hạng này cho CHÍNH entity không? Cùng cửa sổ ±40 với §1.7. */
function proseAwardsTier(e: OcopEntityLike | null | undefined, tier: number): boolean {
  const prose = `${e?.summary || ''} ${e?.description || ''}`
  const needle = new RegExp(`${tier}\\s*sao`, 'gi')
  let m: RegExpExecArray | null
  while ((m = needle.exec(prose)) !== null) {
    const window = prose.slice(Math.max(0, m.index - CLAIM_WINDOW), m.index + m[0].length + CLAIM_WINDOW)
    if (AWARDED.test(window)) return true
  }
  return false
}

/**
 * Con số trong ô OCOP là RÁC CHÉP NHẦM CỘT, không phải một chứng nhận.
 *
 * Đo trên `web/data.json` 2026-08-30: 13 cơ sở lưu trú mang `ocop_star`, và CẢ
 * 13 có `ocop_star` bằng đúng `star_rating` của chính nó (1=1, 2=2, 4=4, 5=5),
 * `ocop_certified` rỗng. Đó là vân tay của một lượt nhập chép nhầm cột
 * hạng-sao-khách-sạn sang ô OCOP. Hậu quả trên trang: một khách sạn hiện
 * «Sản phẩm OCOP 1 sao — Chương trình Mỗi xã Một sản phẩm». Khai khống một
 * chứng nhận NHÀ NƯỚC là đúng thứ CLAUDE.md §1.7 cấm.
 *
 * HAI LUẬT, mỗi luật tự đứng được:
 *  (A) Hạng dưới 3 — OCOP không có bậc đó. Bắt 12 entity (11 khách sạn +
 *      `khu-du-lich-truong-an`). Toàn kho KHÔNG có sản phẩm nào mang hạng 1|2,
 *      nên luật này không đụng một sản phẩm thật nào.
 *  (B) Số chỉ chép lại `star_rating` VÀ văn xuôi không xác nhận hạng đó. Bắt nốt
 *      `homestay-sokfram` (5=5; văn xuôi chỉ nói nó BÀY BÁN sản phẩm OCOP 3–5
 *      sao của địa phương — cùng bẫy "danh mục của người khác" mà
 *      `dua-sap-cau-ke` đã dạy).
 *
 * VÌ SAO (B) phải có vế văn xuôi: `somo-farm-cuu-long` cũng 4=4 nhưng ghi rõ
 * "Đạt chứng nhận OCOP 4 sao năm 2023 cho sản phẩm du lịch sinh thái" — OCOP
 * nhóm 6 (dịch vụ du lịch) là CÓ THẬT. Thiếu vế đó là bóp mất một chứng nhận
 * đúng. Đã đối chiếu từng ca trong cả 13.
 *
 * KHÔNG đụng `ocop_star: 9|0|true` — chúng giữ hành vi cũ (không rút được hạng
 * nhưng vẫn tính là có dấu hiệu): đó là ý định ghi OCOP kèm lỗi gõ, khác hẳn
 * việc chép nhầm nguyên một cột khác.
 */
export function ocopNumberIsColumnBleed(e: OcopEntityLike | null | undefined): boolean {
  const a = attrs(e)
  const n = numericKeyTier(a)
  if (n <= 0) return false
  if (n < MIN_GRADE) return true
  const lodgingStars = a.star_rating
  if (lodgingStars === null || lodgingStars === undefined || lodgingStars === '') return false
  const parsed = parseInt(String(lodgingStars).trim(), 10)
  if (!Number.isFinite(parsed) || parsed !== n) return false
  return !proseAwardsTier(e, n)
}

/**
 * Hạng ghi trong dữ liệu có phải hạng ĐÃ ĐẠT không, hay mới chỉ được đề nghị?
 *
 * Luật HAI CHIỀU, không phải một. Luật một chiều ("văn xuôi có chữ đề xuất thì
 * hạ hạng") đánh oan ngay ca thật đầu tiên gặp phải:
 *
 *   khoai-lang-say-binh-tan  ocop_star=5  "…được ĐỀ XUẤT lên Trung ương đánh giá
 *                                          5 sao OCOP…"        → 5 là CHƯA đạt
 *   khoai-lang-say-dong-phat ocop_star=4  "…ĐẠT OCOP 4 sao và được đề xuất công
 *                                          nhận 5 sao."        → 4 CÓ THẬT
 *
 * Cả hai đều chứa "đề xuất" trong cửa sổ ±40 quanh hạng của mình. Cái phân biệt
 * chúng là ca thứ hai có chữ ĐẠT ngay trước hạng. Nên: chỉ hạ khi cửa sổ có lời
 * đề nghị mà KHÔNG có lời xác nhận. Cùng lớp bài học với cửa sổ ±40 của
 * `_has_stale_geography` bên backend — chuỗi trần giết oan cách viết đúng chuẩn.
 */
export function ocopStarProvisional(e: OcopEntityLike | null | undefined): boolean {
  return proseProposesTier(e, ocopClaimedStars(e))
}

/** Cùng luật, nhưng xét một hạng CHO TRƯỚC — cần vì `ocopStars` có thể đang xét
 *  hạng lấy từ ô văn xuôi sau khi loại ô số rác. */
function proseProposesTier(e: OcopEntityLike | null | undefined, claimed: number): boolean {
  if (claimed <= 0) return false
  const prose = `${e?.summary || ''} ${e?.description || ''}`
  if (!prose.trim()) return false

  const needle = new RegExp(`${claimed}\\s*sao`, 'gi')
  let sawProposal = false
  let sawAwarded = false
  let m: RegExpExecArray | null
  while ((m = needle.exec(prose)) !== null) {
    const window = prose.slice(
      Math.max(0, m.index - CLAIM_WINDOW),
      m.index + m[0].length + CLAIM_WINDOW,
    )
    if (PROPOSAL.test(window)) sawProposal = true
    if (AWARDED.test(window)) sawAwarded = true
  }
  // Xét TOÀN VĂN, không xét từng lần nhắc. Văn xuôi của một entity tự mâu thuẫn
  // với chính nó là chuyện thật, không phải giả định: `khoai-lang-say-binh-tan`
  // nói "được đề xuất lên Trung ương đánh giá 5 sao" HAI lần rồi kết bằng câu
  // trần "Sản phẩm OCOP 5 sao." Luật xét-từng-lần để đúng câu trần đó lật ngược
  // cả phán quyết và trang lại khai 5 sao. Cùng họ với bẫy sáu-ô §5c: một sự
  // thật nằm ở nhiều ô, các ô nói ngược nhau.
  //
  // Khi mâu thuẫn thì lời ĐỀ NGHỊ thắng lời khẳng định trần — nó cụ thể hơn, và
  // §1.7 buộc nghiêng về phía KHÔNG khai khống. Chỉ lời xác nhận TƯỜNG MINH
  // ("đạt N sao", "cấp quốc gia") mới lật lại được.
  return sawProposal && !sawAwarded
}

/** Hạng DÙNG ĐƯỢC để xếp bậc và trưng ra. 0 = có chứng nhận nhưng chưa rõ hạng.
 *
 * KẸP về thang 1..5 ở ĐÂY, không phải ở nơi hiển thị. Bản đầu chỉ kẹp trong
 * `ocopBadgeLabel` nên `ocop_star: 9` vẫn ra tier 9 cho mọi nơi gọi khác — và
 * bộ ca dùng chung với bản Python (`tests/fixtures/ocop-twin-cases.json`) bắt
 * được ngay: Python trả 0, TS trả 9.
 */
export function ocopStars(e: OcopEntityLike | null | undefined): number {
  // Ô số là rác chép nhầm cột → bỏ nó, chỉ còn ô văn xuôi được nói.
  const n = ocopNumberIsColumnBleed(e) ? proseFieldTier(attrs(e)) : ocopClaimedStars(e)
  if (n <= 0) return 0
  if (proseProposesTier(e, n)) return 0
  return n >= 1 && n <= 5 ? n : 0
}

/** Có dấu hiệu OCOP nào không — kể cả khi không rút được hạng. */
export function isOcopCertified(e: OcopEntityLike | null | undefined): boolean {
  const a = attrs(e)
  const hasNumber = NUMERIC_KEYS.some(k => a[k] !== undefined && a[k] !== null && a[k] !== '')
  // Ô số rác chép nhầm cột KHÔNG phải dấu hiệu OCOP — nếu tính, 12 khách sạn vẫn
  // đeo huy hiệu «OCOP» trơn, vẫn lọt bộ lọc và sổ vinh danh.
  if (hasNumber && !ocopNumberIsColumnBleed(e)) return true
  return !!(a.ocop || a.ocop_certified)
}

/**
 * Nhãn hiển thị cho huy hiệu OCOP — «OCOP 5 sao» · «OCOP» · '' (không có).
 *
 * TUYỆT ĐỐI không in `attributes.ocop` thô ra giao diện. Trường đó là văn xuôi
 * tự do, và ít nhất một entity mang cả danh mục sản phẩm của công ty KHÁC:
 *   dua-sap-cau-ke → "VICOSAP: 4 SP OCOP 5 sao quốc gia + 7 SP OCOP 4 sao"
 * In thô ra thì trang gán chứng nhận của người khác cho trái dừa — đo được
 * 2026-08-27: huy hiệu rộng 468px, và chuỗi đó còn lọt vào JSON-LD `brand.name`
 * cho Google đọc. Một entity khác mang số quyết định:
 *   "OCOP 4 sao (QĐ 114/QĐ-UBND, 15/1/2020)"
 */
export function ocopBadgeLabel(e: OcopEntityLike | null | undefined): string {
  if (!isOcopCertified(e)) return ''
  const tier = ocopStars(e)
  return tier > 0 ? `OCOP ${tier} sao` : 'OCOP'
}
